fix: accept 'undefined' as valid value for Prisms, Lens, ISO and Equivalence

arjunvegda opened the pull request.

dai-shi commented on src/focusAtom.ts:
```diff
@@ -49,6 +54,13 @@ export function focusAtom<S, A, R extends void | Promise<void>>(
+  callback: (
+    optic: O.OpticFor<S | undefined>
+  ) => O.Lens<S, any, A> | O.Equivalence<S, any, A> | O.Iso<S, any, A>
+): WritableAtom<A | undefined, SetStateAction<A>, R>
```
curious if you are sure that the return atom value is `A | undefined` instead of `A`.

arjunvegda replied:
Good catch! Fixed it in [this commit](https://github.com/jotai-labs/jotai-optics/pull/5/commits/6976437315d2040d5775f63a0fe06fd894308f17). I think the other way around would be disallowed in `optics-ts` so you're right

dai-shi reviewed (commented):
- one comment below
- another is, is there a reason that `O.Traversal` is not supported?
- last is, we need to support Promise versions too.

Please also be aware that we have #4. It should be fine though.

arjunvegda commented:
> * another is, is there a reason that `O.Traversal` is not supported?

Sorry, I couldn't think of a good example where traversal would be `undefined`. Wouldn't doing so make it a `prism`? Do you have an example of this? My `optics-ts` knowledge is minimal 😄 

> * last is, we need to support Promise versions too.

What would be a good example of this? Could I await for promises inside the `focus` atom? I tinkered with `async` atoms but was not sure if I could retrieve the values to derive an atom.

Here is an example
```ts
type BillingData = {
  id: string
}

type CustomerData = {
  id: string
  billing: BillingData[]
}

const customerListBaseAtom = atom<CustomerData | undefined>(undefined)

const asyncCustomerListAtom = atom(
  (get) => Promise.resolve(get(customerListBaseAtom)),
  async (
    _,
    set,
    nextValue: Promise<CustomerData | undefined>
  ) => {
    set(customerListBaseAtom, await nextValue || undefined)
  }
)
  ```

dai-shi commented:
> My `optics-ts` knowledge is minimal 😄

Me too. Let me try an example.

> > * last is, we need to support Promise versions too.
> 
> What would be a good example of this?

It's simply how jotai works.
```js
const focusedAtom = focusAtom(...)
const asyncAtom = atom(async get () => {
  await something
  return get(focusedAtom)
})
```
So, basically, we should just accept `WritableAtom<Promise<S | undefined>, NonFunction<S>, R>,` in addition to `WritableAtom<S | undefined, NonFunction<S>, R>,`

arjunvegda commented:
> > My `optics-ts` knowledge is minimal 😄
> 
> Me too. Let me try an example.
> 
> > > * last is, we need to support Promise versions too.
> > 
> > 
> > What would be a good example of this?
> 
> It's simply how jotai works.
> 
> ```js
> const focusedAtom = focusAtom(...)
> const asyncAtom = atom(async get () => {
>   await something
>   return get(focusedAtom)
> })
> ```
> 
> So, basically, we should just accept `WritableAtom<Promise<S | undefined>, NonFunction<S>, R>,` in addition to `WritableAtom<S | undefined, NonFunction<S>, R>,`

`Promise<S | undefined>` should be covered by `WritableAtom<S, NonFunction<S>, R>`, no? Since the promise would narrow to `S`?

And `Promise<S | undefined> | undefined` should be covered by [the new type](https://github.com/jotai-labs/jotai-optics/blob/fix-types/src/focusAtom.ts#L40-L43) we added in this PR. I added a [test case for it, to be safe](https://github.com/jotai-labs/jotai-optics/blob/fbfe66e05f69d998cc2f1951c8e8da51a37b0431/src/focusAtom.ts#L40-L43).

dai-shi commented:
> > My `optics-ts` knowledge is minimal 😄
> 
> Me too. Let me try an example.

Here's an example in which `elems()` works for `S | undefined`.
https://codesandbox.io/s/jolly-austin-hdnctw?file=/src/App.tsx

dai-shi commented:
> it covers, but doesn't remove `| undefined`, no?

Hi, I modified the test in 14250af, which I think we should support. I didn't add the fix but confirmed locally that it can be fixed.
We also need the same test for focus-lens.test.tsx and focus-traversal.test.tsx. Let me know if I should work on them.

arjunvegda commented:
> > it covers, but doesn't remove `| undefined`, no?
> 
> Hi, I modified the test in [14250af](https://github.com/jotai-labs/jotai-optics/commit/14250afc77a28f7616700972cf974adfc05753cc), which I think we should support. I didn't add the fix but confirmed locally that it can be fixed. We also need the same test for focus-lens.test.tsx and focus-traversal.test.tsx. Let me know if I should work on them.

Oh wow, I didn’t even know we could chain async atoms like that. I’d have to rely on your expertise here. Could pls you help add the fix whenever you get a chance? 😅

dai-shi commented:
Here we go!

arjunvegda commented on __tests__/focus-traversal.test.tsx:
```diff
@@ -34,3 +34,42 @@ it('updates traversals', async () => {
+  someOtherData: string
+}
+
+it('typescript should accept "undefined" as valid value for traversals', async () => {
+  const customerListListAtom = atom<CustomerData[][]>([])
+
+  const nonEmptyCustomerListAtom = focusAtom(customerListListAtom, (optic) =>
+    optic.find((el) => el.length > 0)
+  )
+
+  focusAtom(nonEmptyCustomerListAtom, (optic) => {
+    const result = optic.elems()
+    return result
+  })
+})
+
+it('should work with promise based atoms with "undefined" value', async () => {
+  const customerBaseAtom = atom<CustomerData[] | undefined>(undefined)
+
+  const asyncCustomerDataAtom = atom(
+    async (get) => get(customerBaseAtom),
+    (_, set, nextValue: CustomerData[]) => {
+      set(customerBaseAtom, nextValue)
+    }
+  )
+
+  focusAtom(asyncCustomerDataAtom, (optic) => {
+    const result = optic.elems()
+    return result
+  })
+})
```
@dai-shi I think these tests are missing the `expect` assertion, is that intentional?

dai-shi replied:
It comes with an internal type, but let me try again.

dai-shi replied:
44d5ae2 added something, but module resolver complains. Any workaround?

arjunvegda replied:
Seems like Traversals are not designed to support `undefined`. It has zero` or `more` _focuses_ 🤔 so the return type must always be an array.

I found this in [their docs](https://akheron.github.io/optics-ts/reference-mc/#collect).
<img width="791" alt="image" src="https://user-images.githubusercontent.com/14841132/214732455-5102756b-013b-45aa-b158-731aede243e1.png">

So to fix it users would technically need to provide an empty array for undefined values?

Something like this?

```diff
it('should work with promise based atoms with "undefined" value', async () => {
  ...
  const focusedPromiseAtom = focusAtom(asyncCustomerDataAtom, (optic) => {
-     const result = optic.elems()
+    const result = optic.valueOr([] as CustomerData[]).elems()
      return result
  })

  expectTypeOf(focusedPromiseAtom).toMatchTypeOf<
-      NotAnArrayType<CustomerData[] | undefined>[],
-      SetStateAction<NotAnArrayType<CustomerData[] | undefined>>,
-      void
+    WritableAtom<CustomerData[], SetStateAction<CustomerData>, void>
  >()
})
```

What do you think?

dai-shi replied:
Sounds reasonable!
Now, I'm pretty confident that you understand optics-ts better than I do. 😄 

Fixed in 729e174.

arjunvegda replied:
haha, nice!! glad we made it work!

Could you help me release this too?

dai-shi replied:
Sure, I will release it.

arjunvegda commented on __tests__/focus-prism.test.tsx:
```diff
@@ -65,3 +67,53 @@ it('atoms that focus on no values are not updated', async () => {
+    optic.optional()
+  )
+
+  expectTypeOf(focusedPromiseAtom).toMatchTypeOf<
+    WritableAtom<Promise<string | undefined>, SetStateAction<CustomerData>>
+  >
```
@dai-shi shouldn't this fail too? 🤔 

The Type would be narrowed to `WritableAtom<CustomerData | undefined, SetStateAction<CustomerData>, void>`, no?

dai-shi replied:
yeah, something was totally wrong. 😓

dai-shi commented on __tests__/focus-lens.test.tsx:
```diff
@@ -216,3 +217,54 @@ it('basic derivation using focus with scope works', async () => {
+    return result
+  })
+
+  expectTypeOf(focusedPromiseAtom).toMatchTypeOf<
+    WritableAtom<Promise<string>, SetStateAction<CustomerData>>
+  >
```
Oops, sorry about that.
You are right. (In #4, it will be `Promise<string>` instead of `string`.)
